docker-compose.yml

```diff
@@ -0,0 +1,43 @@
+version: "3.3"
+
+services:
+  database:
+    image: mysql
+    restart: on-failure
+    ports:
+      - 3306:3306
+    expose:
+      - 3306
+    environment:
+      - MYSQL_DATABASE=hollowverse
+      - MYSQL_ROOT_PASSWORD=123456
+      - MYSQL_ROOT_HOST=%
+
+  api:
+    restart: always
+    depends_on:
+      - database
+    build:
+      dockerfile: Dockerfile-dev
+      context: .
+    links:
+      - database
+    ports:
+      - 8080:8080
+      - 9229:9229
+    volumes:
+      - type: bind
+        source: .
+        target: /code
+      
+      - type: volume
```

Comment:
What's the difference between type `bind` and type `volume`?

Reply:
`bind` tells Docker to use an existing directory on the host OS as a volume inside the container. `volume` creates an empty volume, managed by Docker, so that the data can be persisted even if the container is stopped. We need `bind` to mount the project code and allow automatic reloading on code changes. `volume` is used to cache `node_modules` separately from the host `node_modules` so that dependencies do not conflict (suppose that the host is Windows, if the container overwrites `node_modules`, some native dependencies will not work when running the server outside the container)

Reply:
The volume created with type `volume` can be inspected like this:

```
❯ docker volume inspect api_node_modules
[
    {
        "CreatedAt": "2017-10-30T07:23:24+02:00",
        "Driver": "local",
        "Labels": {
            "com.docker.compose.project": "api",
            "com.docker.compose.volume": "node_modules"
        },
        "Mountpoint": "/var/lib/docker/volumes/api_node_modules/_data",
        "Name": "api_node_modules",
        "Options": {},
        "Scope": "local"
    }
]

```